File: core/src/main/kotlin/com/backend/core/auth/AuthService.kt
package com.backend.core.auth

import com.backend.core.security.JwtService
import com.backend.core.security.TokenPair
import com.backend.core.user.UserEntity
import com.backend.core.user.UserRepository
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.*

@Service
@Transactional
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService,
    private val authenticationManager: AuthenticationManager,
    private val twoFactorService: TwoFactorService
) {

    private val objectMapper = jacksonObjectMapper()

    fun login(email: String, password: String, twoFactorCode: Int? = null, tenantId: String? = null): TokenPair {
        try {
            // Step 1: Authenticate user credentials
            authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(email, password)
            )

            // Step 2: Get user details
            val user = userRepository.findActiveByEmail(email)
                ?: throw BadCredentialsException("Invalid email or password")

            // Step 3: Check if account is locked
            if (user.lockedUntil?.isAfter(LocalDateTime.now()) == true) {
                throw BadCredentialsException("Account is temporarily locked")
            }

            // Step 4: Check 2FA if enabled
            if (user.twoFactorEnabled) {
                if (twoFactorCode == null) {
                    throw TwoFactorRequiredException("Two-factor authentication code required")
                }

                // Verify 2FA code
                val isValidCode = user.twoFactorSecret?.let { secret ->
                    twoFactorService.verifyCode(secret, twoFactorCode)
                } ?: false

                // Check backup codes if main code fails
                val isValidBackupCode = if (!isValidCode && user.backupCodes != null) {
                    verifyBackupCode(user, twoFactorCode.toString())
                } else false

                if (!isValidCode && !isValidBackupCode) {
                    throw BadCredentialsException("Invalid two-factor authentication code")
                }

                // Update 2FA verification timestamp
                val updatedUser = user.copy(twoFactorVerifiedAt = LocalDateTime.now())
                userRepository.save(updatedUser)
            }

            // Step 5: Reset failed login attempts and update last login
            val updatedUser = user.copy(
                failedLoginAttempts = 0,
                lastLogin = LocalDateTime.now()
            )
            userRepository.save(updatedUser)

            // Step 6: Generate tokens
            val resolvedTenantId = tenantId ?: "default-tenant"
            val role = if (user.isSuperuser) "admin" else "user"

            return jwtService.generateTokenPair(user, resolvedTenantId, role)

        } catch (e: TwoFactorRequiredException) {
            throw e
        } catch (e: BadCredentialsException) {
            handleFailedLogin(email)
            throw BadCredentialsException("Invalid email or password")
        }
    }

    fun enableTwoFactor(userId: UUID): Map<String, Any> {
        val user = userRepository.findById(userId).orElse(null)
            ?: throw IllegalArgumentException("User not found")

        if (user.twoFactorEnabled) {
            throw IllegalArgumentException("Two-factor authentication is already enabled")
        }

        val secretKey = twoFactorService.generateSecretKey()
        val qrCodeUrl = twoFactorService.generateQRCodeUrl(user.email, secretKey)
        val backupCodes = twoFactorService.generateBackupCodes()

        val updatedUser = user.copy(
            twoFactorSecret = secretKey,
            backupCodes = objectMapper.writeValueAsString(backupCodes),
            updatedAt = LocalDateTime.now()
        )
        userRepository.save(updatedUser)

        return mapOf(
            "secretKey" to secretKey,
            "qrCodeUrl" to qrCodeUrl,
            "backupCodes" to backupCodes,
            "message" to "Scan QR code with your authenticator app and verify with a code to complete setup"
        )
    }

    fun verifyAndEnableTwoFactor(userId: UUID, verificationCode: Int): Map<String, String> {
        val user = userRepository.findById(userId).orElse(null)
            ?: throw IllegalArgumentException("User not found")

        if (user.twoFactorEnabled) {
            throw IllegalArgumentException("Two-factor authentication is already enabled")
        }

        val secretKey = user.twoFactorSecret
            ?: throw IllegalArgumentException("Two-factor setup not initiated")

        if (!twoFactorService.verifyCode(secretKey, verificationCode)) {
            throw IllegalArgumentException("Invalid verification code")
        }

        val updatedUser = user.copy(
            twoFactorEnabled = true,
            twoFactorVerifiedAt = LocalDateTime.now(),
            updatedAt = LocalDateTime.now()
        )
        userRepository.save(updatedUser)

        return mapOf("message" to "Two-factor authentication enabled successfully")
    }

    fun disableTwoFactor(userId: UUID, verificationCode: Int): Map<String, String> {
        val user = userRepository.findById(userId).orElse(null)
            ?: throw IllegalArgumentException("User not found")

        if (!user.twoFactorEnabled) {
            throw IllegalArgumentException("Two-factor authentication is not enabled")
        }

        val secretKey = user.twoFactorSecret
            ?: throw IllegalArgumentException("Two-factor secret not found")

        if (!twoFactorService.verifyCode(secretKey, verificationCode)) {
            throw IllegalArgumentException("Invalid verification code")
        }

        val updatedUser = user.copy(
            twoFactorEnabled = false,
            twoFactorSecret = null,
            backupCodes = null,
            twoFactorVerifiedAt = null,
            updatedAt = LocalDateTime.now()
        )
        userRepository.save(updatedUser)

        return mapOf("message" to "Two-factor authentication disabled successfully")
    }

    fun regenerateBackupCodes(userId: UUID, verificationCode: Int): Map<String, Any> {
        val user = userRepository.findById(userId).orElse(null)
            ?: throw IllegalArgumentException("User not found")

        if (!user.twoFactorEnabled) {
            throw IllegalArgumentException("Two-factor authentication is not enabled")
        }

        val secretKey = user.twoFactorSecret
            ?: throw IllegalArgumentException("Two-factor secret not found")

        if (!twoFactorService.verifyCode(secretKey, verificationCode)) {
            throw IllegalArgumentException("Invalid verification code")
        }

        val newBackupCodes = twoFactorService.generateBackupCodes()

        val updatedUser = user.copy(
            backupCodes = objectMapper.writeValueAsString(newBackupCodes),
            updatedAt = LocalDateTime.now()
        )
        userRepository.save(updatedUser)

        return mapOf(
            "backupCodes" to newBackupCodes,
            "message" to "New backup codes generated. Store them safely!"
        )
    }

    private fun verifyBackupCode(user: UserEntity, code: String): Boolean {
        val backupCodesJson = user.backupCodes ?: return false
        val backupCodes = objectMapper.readValue(backupCodesJson, Array<String>::class.java).toMutableList()

        return if (backupCodes.contains(code.uppercase())) {
            // Remove used backup code
            backupCodes.remove(code.uppercase())
            val updatedUser = user.copy(
                backupCodes = objectMapper.writeValueAsString(backupCodes),
                updatedAt = LocalDateTime.now()
            )
            userRepository.save(updatedUser)
            true
        } else {
            false
        }
    }

    private fun handleFailedLogin(email: String) {
        val user = userRepository.findByEmail(email) ?: return
        val attempts = user.failedLoginAttempts + 1
        val updatedUser = if (attempts >= 5) {
            user.copy(
                failedLoginAttempts = attempts,
                lockedUntil = LocalDateTime.now().plusMinutes(30)
            )
        } else {
            user.copy(failedLoginAttempts = attempts)
        }
        userRepository.save(updatedUser)
    }

    // ... rest of existing methods (register, refreshToken, changePassword, etc.)
    fun register(email: String, password: String, passwordConfirm: String, firstName: String, lastName: String): Map<String, Any> {
        if (password != passwordConfirm) {
            throw IllegalArgumentException("Passwords do not match")
        }

        if (userRepository.existsByEmail(email)) {
            throw IllegalArgumentException("User with email $email already exists")
        }

        validatePasswordStrength(password)

        val user = UserEntity(
            email = email,
            hashedPassword = passwordEncoder.encode(password),
            firstName = firstName,
            lastName = lastName,
            isActive = true,
            isVerified = true
        )

        val savedUser = userRepository.save(user)

        return mapOf(
            "message" to "Registration successful. You can now log in.",
            "userId" to savedUser.id.toString(),
            "email" to savedUser.email
        )
    }

    fun refreshToken(refreshToken: String): TokenPair {
        try {
            if (!jwtService.isRefreshToken(refreshToken)) {
                throw BadCredentialsException("Invalid token type")
            }

            val userEmail = jwtService.extractUsername(refreshToken)
            val user = userRepository.findActiveByEmail(userEmail)
                ?: throw BadCredentialsException("User not found")

            if (jwtService.isTokenValid(refreshToken, user)) {
                val tenantId = jwtService.extractTenantId(refreshToken)
                val role = jwtService.extractRole(refreshToken)

                return jwtService.generateTokenPair(user, tenantId, role)
            } else {
                throw BadCredentialsException("Invalid or expired token")
            }
        } catch (e: Exception) {
            throw BadCredentialsException("Failed to refresh token")
        }
    }

    fun changePassword(userId: UUID, currentPassword: String, newPassword: String, newPasswordConfirm: String) {
        if (newPassword != newPasswordConfirm) {
            throw IllegalArgumentException("New passwords do not match")
        }

        val user = userRepository.findById(userId).orElse(null)
            ?: throw IllegalArgumentException("User not found")

        if (!passwordEncoder.matches(currentPassword, user.hashedPassword)) {
            throw IllegalArgumentException("Current password is incorrect")
        }

        validatePasswordStrength(newPassword)

        if (passwordEncoder.matches(newPassword, user.hashedPassword)) {
            throw IllegalArgumentException("New password must be different from current password")
        }

        val updatedUser = user.copy(
            hashedPassword = passwordEncoder.encode(newPassword),
            failedLoginAttempts = 0
        )
        userRepository.save(updatedUser)
    }

    fun getCurrentUser(userId: UUID): UserEntity {
        return userRepository.findById(userId).orElse(null)
            ?: throw IllegalArgumentException("User not found")
    }

    private fun validatePasswordStrength(password: String) {
        val minLength = 8
        if (password.length < minLength) {
            throw IllegalArgumentException("Password must be at least $minLength characters long")
        }
        if (!password.any { it.isDigit() }) {
            throw IllegalArgumentException("Password must contain at least one digit")
        }
        if (!password.any { it.isUpperCase() }) {
            throw IllegalArgumentException("Password must contain at least one uppercase letter")
        }
        if (!password.any { it.isLowerCase() }) {
            throw IllegalArgumentException("Password must contain at least one lowercase letter")
        }
        if (!password.any { "!@#$%^&*()_+-=[]{}|;:,.<>?".contains(it) }) {
            throw IllegalArgumentException("Password must contain at least one special character")
        }
    }

    fun getRolePermissions(role: String): List<String> {
        return when (role) {
            "admin" -> listOf(
                "read:all", "write:all", "delete:all",
                "manage:users", "manage:settings"
            )
            "manager" -> listOf(
                "read:contacts", "write:contacts", "read:companies",
                "write:companies", "read:deals", "write:deals"
            )
            "user" -> listOf(
                "read:contacts", "write:contacts:own", "read:companies",
                "read:deals", "write:deals:own"
            )
            else -> emptyList()
        }
    }
}

File: identity-service/src/auth/auth.service.ts
import {
  BadRequestException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import * as bcrypt from 'bcryptjs';
import { TwoFactorService } from '../two-factor/two-factor.service';
import { UserService } from '../user/user.service';
import { JwtService } from './jwt.service';
import {
  AccountLockedException,
  AuthUser,
  InvalidCredentialsException,
  LoginDto,
  RefreshTokenDto,
  RegisterDto,
  TokenPair,
  TwoFactorRequiredException,
} from './types';

@Injectable()
export class AuthService {
  constructor(
    private readonly userService: UserService,
    private readonly jwtService: JwtService,
    private readonly twoFactorService: TwoFactorService,
  ) {}

  async login(loginDto: LoginDto): Promise<TokenPair> {
    const { email, password, twoFactorCode, tenantId } = loginDto;

    try {
      console.log(`Login attempt for email: ${email}`);

      // Step 1: Find and validate user
      const user = await this.userService.findActiveByEmail(email);
      if (!user) {
        console.log(`User not found for email: ${email}`);
        throw new InvalidCredentialsException('Invalid email or password');
      }

      console.log(
        `User found: ${user.email}, isActive: ${user.isActive}, isVerified: ${user.isVerified}`,
      );

      // Step 2: Check if account is locked
      if (user.lockedUntil && user.lockedUntil > new Date()) {
        console.log(`Account locked until: ${user.lockedUntil}`);
        throw new AccountLockedException('Account is temporarily locked');
      }

      // Step 3: Verify password
      console.log(`Verifying password for user: ${user.email}`);
      const isPasswordValid = await bcrypt.compare(
        password,
        user.hashedPassword,
      );
      console.log(`Password valid: ${isPasswordValid}`);

      if (!isPasswordValid) {
        console.log(`Invalid password for user: ${email}`);
        // Increment failed login attempts
        await this.userService.incrementFailedLoginAttempts(user.id);
        throw new InvalidCredentialsException('Invalid email or password');
      }

      // Step 4: Check 2FA if enabled
      if (user.twoFactorEnabled) {
        console.log(`2FA enabled for user: ${email}`);
        if (!twoFactorCode) {
          throw new TwoFactorRequiredException(
            'Two-factor authentication code required',
          );
        }

        // Verify 2FA code using TwoFactorService
        console.log(
          `Raw 2FA code received: ${twoFactorCode}, type: ${typeof twoFactorCode}`,
        );

        // Pass the code as-is to TwoFactorService
        const isValid2FA = await this.twoFactorService.verifyTwoFactorCode(
          user.id,
          twoFactorCode,
        );
        if (!isValid2FA) {
          console.log(`Invalid 2FA code for user: ${email}`);
          throw new InvalidCredentialsException(
            'Invalid two-factor authentication code',
          );
        }

        console.log(`2FA verification successful for user: ${email}`);
      }

      // Step 5: Reset failed login attempts and update last login
      await this.userService.resetFailedLoginAttempts(user.id);

      // Step 6: Generate and return tokens
      const resolvedTenantId = tenantId || 'default-tenant';
      const role = user.isSuperuser ? 'admin' : 'user';

      console.log(
        `Generating tokens for user: ${email}, tenant: ${resolvedTenantId}, role: ${role}`,
      );

      const tokenPair = this.jwtService.generateTokenPair(
        user,
        resolvedTenantId,
        role,
      );
      console.log(`Login successful for user: ${email}`);

      return tokenPair;
    } catch (error) {
      console.error(`Login error for ${email}:`, error.message);

      if (
        error instanceof TwoFactorRequiredException ||
        error instanceof InvalidCredentialsException ||
        error instanceof AccountLockedException
      ) {
        throw error;
      }

      console.error(`Unexpected login error:`, error);
      throw new UnauthorizedException('Login failed');
    }
  }

  async register(
    registerDto: RegisterDto,
  ): Promise<{ message: string; userId: string; email: string }> {
    const { email, password, passwordConfirm, firstName, lastName } =
      registerDto;

    try {
      console.log(`Registration attempt for email: ${email}`);

      // Validate password confirmation
      if (password !== passwordConfirm) {
        console.log('Password confirmation mismatch');
        throw new BadRequestException('Passwords do not match');
      }

      // Validate password strength
      console.log('Validating password strength...');
      this.validatePasswordStrength(password);

      // Check if user already exists
      console.log(`Checking if user exists: ${email}`);
      const existingUser = await this.userService.findByEmail(email);
      if (existingUser) {
        console.log(`User already exists: ${email}`);
        throw new BadRequestException(
          `User with email ${email} already exists`,
        );
      }

      // Create user using UserService
      console.log(`Creating user: ${email}`);
      const user = await this.userService.create({
        email,
        password, // UserService will handle hashing
        firstName,
        lastName,
      });

      console.log(
        `User created successfully: ${user.email} with ID: ${user.id}`,
      );

      return {
        message: 'Registration successful. You can now log in.',
        userId: user.id,
        email: user.email,
      };
    } catch (error) {
      console.error(`Registration error for ${email}:`, error.message);

      if (error instanceof BadRequestException) {
        throw error;
      }

      console.error(`Unexpected registration error:`, error);
      throw new BadRequestException('Registration failed');
    }
  }

  async refreshToken(refreshTokenDto: RefreshTokenDto): Promise<TokenPair> {
    const { refreshToken } = refreshTokenDto;

    try {
      console.log(
        `Refresh token attempt with token: ${refreshToken.substring(0, 20)}...`,
      );

      // Verify refresh token
      const payload = this.jwtService.verifyRefreshToken(refreshToken);
      console.log(`Token payload:`, {
        sub: payload.sub,
        tenantId: payload.tenantId,
        role: payload.role,
        type: payload.type,
      });

      // Find user by email (since sub contains email in our implementation)
      const user = await this.userService.findByEmail(payload.sub);
      if (!user) {
        console.log(`User not found for email: ${payload.sub}`);
        throw new UnauthorizedException('User not found');
      }

      if (!user.isActive) {
        console.log(`User is not active: ${payload.sub}`);
        throw new UnauthorizedException('User not found or inactive');
      }

      console.log(`User found: ${user.email}, isActive: ${user.isActive}`);

      // Validate refresh token against user
      if (!this.jwtService.isRefreshTokenValid(refreshToken, user)) {
        console.log(`Invalid refresh token for user: ${user.email}`);
        throw new UnauthorizedException('Invalid refresh token');
      }

      // Generate new token pair
      console.log(
        `Generating new token pair for user: ${user.email}, tenant: ${payload.tenantId}, role: ${payload.role}`,
      );
      const newTokenPair = this.jwtService.generateTokenPair(
        user,
        payload.tenantId,
        payload.role,
      );

      console.log(`Refresh successful for user: ${user.email}`);
      return newTokenPair;
    } catch (error) {
      console.error(`Refresh token error:`, error.message);

      if (error instanceof UnauthorizedException) {
        throw error;
      }

      console.error(`Unexpected refresh error:`, error);
      throw new UnauthorizedException('Failed to refresh token');
    }
  }

  async getCurrentUser(userId: string): Promise<AuthUser> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('User not found');
    }

    return {
      id: user.id,
      email: user.email,
      firstName: user.firstName,
      lastName: user.lastName,
      isActive: user.isActive,
      isVerified: user.isVerified,
      isSuperuser: user.isSuperuser,
      authorities: user.isSuperuser
        ? ['ROLE_ADMIN', 'ROLE_USER']
        : ['ROLE_USER'],
    };
  }

  async logout(userId: string): Promise<{ message: string }> {
    // TODO: Implement token blacklisting with Redis
    // For now, we'll just return a success message
    // The client should discard the tokens
    console.log(`User ${userId} logged out`);

    return {
      message: 'Logged out successfully',
    };
  }

  async changePassword(
    userId: string,
    currentPassword: string,
    newPassword: string,
    newPasswordConfirm: string,
  ): Promise<void> {
    // Delegate to UserService
    await this.userService.changePassword(userId, {
      currentPassword,
      newPassword,
      newPasswordConfirm,
    });
  }

  validatePasswordStrength(password: string): void {
    const minLength = 8;
    if (password.length < minLength) {
      throw new BadRequestException(
        `Password must be at least ${minLength} characters long`,
      );
    }
    if (!password.match(/\d/)) {
      throw new BadRequestException('Password must contain at least one digit');
    }
    if (!password.match(/[A-Z]/)) {
      throw new BadRequestException(
        'Password must contain at least one uppercase letter',
      );
    }
    if (!password.match(/[a-z]/)) {
      throw new BadRequestException(
        'Password must contain at least one lowercase letter',
      );
    }
    if (!password.match(/[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/)) {
      throw new BadRequestException(
        'Password must contain at least one special character',
      );
    }
  }
}
